Solemne 3: Construir una red neuronal para el conjunto de datos entregados, que contemple los siguientes pasos:

Cargar datos.
Definir modelo.
Modelo de compilación.
Modelo de ajuste.
Evaluar modelo.


Estudiante: José Luis San Martín

Conjunto de datos sobre la diabetes en una población

Describe los datos de los registros médicos de los pacientes  y si tenían la aparición de la diabetes en un plazo de cinco años. Se trata de un problema de clasificación binaria (aparición de diabetes como 1 o no como 0). Las variables de entrada que describen a cada paciente son numéricas y tienen variables de escalas. A continuación se enumeran los ocho atributos del set de datos:

1. Número de embarazos.
2. Concentración de glucosa en plasma a 2 horas en una prueba oral de tolerancia a la glucosa.
3. Presión arterial diastólica (mm Hg).
4. Grosor del pliegue de la piel del tríceps (mm).
5. Insulina sérica de 2 horas (mu U/ml).
6. Índice de masa corporal.
7. Función del pedigrí de la diabetes.
8. Edad (años).
9. Clase, aparición de diabetes en un plazo de cinco años.

In [6]:
import pandas as pd
import numpy as np
from keras.models import Sequential
from keras.layers.core import Dense

Importacion de librerias que se usaran para trabajar en el modelo, dentro de las que destacan librerias de Keras.

In [7]:
caracteristicas_header=['N de embarazos','Glucosa','Presion','Pliegue tricep','Insulina','IMC','Pedigri diabetes','Edad','Clase']
dataFrame=pd.read_csv('pacientes_diabetes.csv',sep=',',header=None, names=caracteristicas_header)
dataFrame.head(10)

N de embarazos  Glucosa  Presion  Pliegue tricep  Insulina   IMC  \
0               6      148       72              35         0  33.6   
1               1       85       66              29         0  26.6   
2               8      183       64               0         0  23.3   
3               1       89       66              23        94  28.1   
4               0      137       40              35       168  43.1   
5               5      116       74               0         0  25.6   
6               3       78       50              32        88  31.0   
7              10      115        0               0         0  35.3   
8               2      197       70              45       543  30.5   
9               8      125       96               0         0   0.0   

   Pedigri diabetes  Edad  Clase  
0             0.627    50      1  
1             0.351    31      0  
2             0.672    32      1  
3             0.167    21      0  
4             2.288    33      1  
5             0.201    30      0  
6             0.248    26      1  
7             0.134    29      0  
8             0.158    53      1  
9             0.232    54      1

Se obtiene el conjunto de datos del socumento csv, el cual estaba sin el encabezado de las caracteristicas. Este se agrego para obtener mayor claridad a la hora de visualizar los datos.

In [8]:
datosParaEntranamiento = np.array(dataFrame.drop(['Clase'],1))

datosDeObjetivo = np.array(dataFrame['Clase'])

In [9]:
model = Sequential()
model.add(Dense(16, input_dim=8, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

Se hara la prueba con un modelo usando 8 entradas y 32 neuronas en la capa oculta

In [10]:
model.compile(loss='mean_squared_error',
              optimizer='adam',
              metrics=['binary_accuracy'])

Presisciones realizadas antes del entrenamiento de la red, estos ajustes son para mostrar el tipo de perdida que se tendra en el modelo. Como el error de media cuadrado, tambien se trabajara con adam y se medira la precision del modelo con binary accuracy

In [11]:
model.fit(datosParaEntranamiento, datosDeObjetivo, epochs=1000)

Epoch 1/1000
24/24 [==============================] - 13s 2ms/step - loss: 0.5038 - binary_accuracy: 0.4690
Epoch 2/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.3761 - binary_accuracy: 0.6072
Epoch 3/1000
24/24 [==============================] - 0s 3ms/step - loss: 0.3115 - binary_accuracy: 0.6800
Epoch 4/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.3498 - binary_accuracy: 0.6440
Epoch 5/1000
24/24 [==============================] - 0s 4ms/step - loss: 0.3327 - binary_accuracy: 0.6574
Epoch 6/1000
24/24 [==============================] - 0s 4ms/step - loss: 0.3368 - binary_accuracy: 0.6545
Epoch 7/1000
24/24 [==============================] - 0s 4ms/step - loss: 0.3409 - binary_accuracy: 0.6519
Epoch 8/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.3360 - binary_accuracy: 0.6544
Epoch 9/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.3248 - binary_accuracy: 0.6723
Epoch 10/1000
24/24 [===============

24/24 [==============================] - 0s 3ms/step - loss: 0.3043 - binary_accuracy: 0.6913
Epoch 77/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.2824 - binary_accuracy: 0.7059
Epoch 78/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.2987 - binary_accuracy: 0.6902
Epoch 79/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.3222 - binary_accuracy: 0.6651
Epoch 80/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.3257 - binary_accuracy: 0.6644
Epoch 81/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.3290 - binary_accuracy: 0.6653
Epoch 82/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.3155 - binary_accuracy: 0.6805
Epoch 83/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.3043 - binary_accuracy: 0.6913
Epoch 84/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.3096 - binary_accuracy: 0.6787
Epoch 85/1000
24/24 [=====================

24/24 [==============================] - 0s 2ms/step - loss: 0.2339 - binary_accuracy: 0.7445
Epoch 152/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.2761 - binary_accuracy: 0.6833
Epoch 153/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.2389 - binary_accuracy: 0.7375
Epoch 154/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.2857 - binary_accuracy: 0.6866
Epoch 155/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.2418 - binary_accuracy: 0.7258
Epoch 156/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.2609 - binary_accuracy: 0.7135
Epoch 157/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.2710 - binary_accuracy: 0.6950
Epoch 158/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.2334 - binary_accuracy: 0.7419
Epoch 159/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.2345 - binary_accuracy: 0.7212
Epoch 160/1000
24/24 [============

24/24 [==============================] - 0s 2ms/step - loss: 0.2496 - binary_accuracy: 0.7121
Epoch 227/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.2448 - binary_accuracy: 0.7326
Epoch 228/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.2831 - binary_accuracy: 0.6571
Epoch 229/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.3183 - binary_accuracy: 0.6620
Epoch 230/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.2603 - binary_accuracy: 0.6869
Epoch 231/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.2547 - binary_accuracy: 0.6766
Epoch 232/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.2260 - binary_accuracy: 0.7368
Epoch 233/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.2070 - binary_accuracy: 0.7521
Epoch 234/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.2312 - binary_accuracy: 0.7166
Epoch 235/1000
24/24 [============

24/24 [==============================] - 0s 2ms/step - loss: 0.1624 - binary_accuracy: 0.7861
Epoch 302/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1843 - binary_accuracy: 0.7368
Epoch 303/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1787 - binary_accuracy: 0.7571
Epoch 304/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1733 - binary_accuracy: 0.7576
Epoch 305/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1638 - binary_accuracy: 0.7717
Epoch 306/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1705 - binary_accuracy: 0.7694
Epoch 307/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1795 - binary_accuracy: 0.7698
Epoch 308/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1682 - binary_accuracy: 0.7706
Epoch 309/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1615 - binary_accuracy: 0.7924
Epoch 310/1000
24/24 [============

24/24 [==============================] - 0s 2ms/step - loss: 0.1539 - binary_accuracy: 0.7802
Epoch 377/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1981 - binary_accuracy: 0.7222
Epoch 378/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1961 - binary_accuracy: 0.7153
Epoch 379/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1708 - binary_accuracy: 0.7622
Epoch 380/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1581 - binary_accuracy: 0.7772
Epoch 381/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1500 - binary_accuracy: 0.7943
Epoch 382/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1777 - binary_accuracy: 0.7494
Epoch 383/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1742 - binary_accuracy: 0.7694
Epoch 384/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1704 - binary_accuracy: 0.7516
Epoch 385/1000
24/24 [============

24/24 [==============================] - 0s 2ms/step - loss: 0.1482 - binary_accuracy: 0.7880
Epoch 452/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1642 - binary_accuracy: 0.7825
Epoch 453/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1561 - binary_accuracy: 0.7763
Epoch 454/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1569 - binary_accuracy: 0.7808
Epoch 455/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1418 - binary_accuracy: 0.8038
Epoch 456/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1534 - binary_accuracy: 0.7922
Epoch 457/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1531 - binary_accuracy: 0.7822
Epoch 458/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1590 - binary_accuracy: 0.7836
Epoch 459/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1627 - binary_accuracy: 0.7769
Epoch 460/1000
24/24 [============

24/24 [==============================] - 0s 2ms/step - loss: 0.1423 - binary_accuracy: 0.8110
Epoch 526/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1468 - binary_accuracy: 0.7976
Epoch 527/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1510 - binary_accuracy: 0.7798
Epoch 528/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1462 - binary_accuracy: 0.8118
Epoch 529/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1466 - binary_accuracy: 0.8056
Epoch 530/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1310 - binary_accuracy: 0.8113
Epoch 531/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1502 - binary_accuracy: 0.8006
Epoch 532/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1646 - binary_accuracy: 0.7732
Epoch 533/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1490 - binary_accuracy: 0.8110
Epoch 534/1000
24/24 [============

24/24 [==============================] - 0s 2ms/step - loss: 0.1598 - binary_accuracy: 0.7714
Epoch 601/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1611 - binary_accuracy: 0.7843
Epoch 602/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1540 - binary_accuracy: 0.7806
Epoch 603/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1622 - binary_accuracy: 0.7682
Epoch 604/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1369 - binary_accuracy: 0.7884
Epoch 605/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1533 - binary_accuracy: 0.7932
Epoch 606/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1596 - binary_accuracy: 0.7803
Epoch 607/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1446 - binary_accuracy: 0.7939
Epoch 608/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1603 - binary_accuracy: 0.7656
Epoch 609/1000
24/24 [============

24/24 [==============================] - 0s 2ms/step - loss: 0.1551 - binary_accuracy: 0.7890
Epoch 676/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1421 - binary_accuracy: 0.8058
Epoch 677/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1586 - binary_accuracy: 0.7885
Epoch 678/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1728 - binary_accuracy: 0.7584
Epoch 679/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1327 - binary_accuracy: 0.8206
Epoch 680/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1512 - binary_accuracy: 0.7882
Epoch 681/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1490 - binary_accuracy: 0.7850
Epoch 682/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1505 - binary_accuracy: 0.7834
Epoch 683/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1724 - binary_accuracy: 0.7535
Epoch 684/1000
24/24 [============

24/24 [==============================] - 0s 2ms/step - loss: 0.1452 - binary_accuracy: 0.8158
Epoch 751/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1594 - binary_accuracy: 0.7923
Epoch 752/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1399 - binary_accuracy: 0.8104
Epoch 753/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1451 - binary_accuracy: 0.7919
Epoch 754/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1530 - binary_accuracy: 0.7902
Epoch 755/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1362 - binary_accuracy: 0.8207
Epoch 756/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1361 - binary_accuracy: 0.8163
Epoch 757/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1457 - binary_accuracy: 0.7944
Epoch 758/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1436 - binary_accuracy: 0.8056
Epoch 759/1000
24/24 [============

24/24 [==============================] - 0s 2ms/step - loss: 0.1526 - binary_accuracy: 0.7991
Epoch 826/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1621 - binary_accuracy: 0.7802
Epoch 827/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1531 - binary_accuracy: 0.7928
Epoch 828/1000
24/24 [==============================] - 0s 4ms/step - loss: 0.1469 - binary_accuracy: 0.8144
Epoch 829/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1484 - binary_accuracy: 0.7764
Epoch 830/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1564 - binary_accuracy: 0.7720
Epoch 831/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1390 - binary_accuracy: 0.7960
Epoch 832/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1658 - binary_accuracy: 0.7574
Epoch 833/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1562 - binary_accuracy: 0.7842
Epoch 834/1000
24/24 [============

24/24 [==============================] - 0s 2ms/step - loss: 0.1354 - binary_accuracy: 0.8106
Epoch 901/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1670 - binary_accuracy: 0.7693
Epoch 902/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1475 - binary_accuracy: 0.7781
Epoch 903/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1308 - binary_accuracy: 0.8216
Epoch 904/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1413 - binary_accuracy: 0.8083
Epoch 905/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1651 - binary_accuracy: 0.7678
Epoch 906/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1435 - binary_accuracy: 0.8003
Epoch 907/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1561 - binary_accuracy: 0.7704
Epoch 908/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1470 - binary_accuracy: 0.8017
Epoch 909/1000
24/24 [============

24/24 [==============================] - 0s 2ms/step - loss: 0.1602 - binary_accuracy: 0.7753
Epoch 976/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1561 - binary_accuracy: 0.7860
Epoch 977/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1435 - binary_accuracy: 0.7950
Epoch 978/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1494 - binary_accuracy: 0.7787
Epoch 979/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1679 - binary_accuracy: 0.7637
Epoch 980/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1802 - binary_accuracy: 0.7489
Epoch 981/1000
24/24 [==============================] - 0s 1ms/step - loss: 0.1631 - binary_accuracy: 0.7659
Epoch 982/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1490 - binary_accuracy: 0.7716
Epoch 983/1000
24/24 [==============================] - 0s 2ms/step - loss: 0.1455 - binary_accuracy: 0.7912
Epoch 984/1000
24/24 [============

se entrena el modelo con una con 1000 iteraciones como parte de aprendizaje de entrenamiento.
En el epoch 1/1000 se obtiene una puntuacion de 0.469, en el siguiente epoch se obtiene una puntiacion de 0.6072 y luego sigue aumentando y con algunos desensos menores en algunos casos. Centrandonos en la primeras filas da entender que el modelo comienza a "aprender", esto debido a el aumento de puntuacion lo que quiere decir que el modelo cada vez comienza a ser mas preciso.

In [12]:
scores = model.evaluate(datosParaEntranamiento, datosDeObjetivo)
print("\n%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))

24/24 [==============================] - 1s 1ms/step - loss: 0.1547 - binary_accuracy: 0.7904

binary_accuracy: 79.04%


In [13]:
print (model.predict(datosParaEntranamiento).round())

[[1.]
 [0.]
 [1.]
 [0.]
 [1.]
 [0.]
 [0.]
 [1.]
 [1.]
 [1.]
 [0.]
 [1.]
 [0.]
 [1.]
 [1.]
 [0.]
 [1.]
 [0.]
 [0.]
 [0.]
 [1.]
 [0.]
 [1.]
 [0.]
 [1.]
 [1.]
 [1.]
 [0.]
 [0.]
 [0.]
 [0.]
 [1.]
 [0.]
 [0.]
 [0.]
 [1.]
 [1.]
 [1.]
 [0.]
 [1.]
 [0.]
 [1.]
 [0.]
 [1.]
 [1.]
 [1.]
 [1.]
 [0.]
 [1.]
 [0.]
 [0.]
 [0.]
 [0.]
 [1.]
 [1.]
 [0.]
 [1.]
 [0.]
 [1.]
 [0.]
 [0.]
 [1.]
 [0.]
 [1.]
 [1.]
 [0.]
 [0.]
 [1.]
 [0.]
 [0.]
 [0.]
 [1.]
 [1.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [1.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [1.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [1.]
 [0.]
 [0.]
 [0.]
 [1.]
 [1.]
 [0.]
 [0.]
 [0.]
 [0.]
 [1.]
 [0.]
 [1.]
 [0.]
 [0.]
 [1.]
 [1.]
 [0.]
 [0.]
 [1.]
 [1.]
 [0.]
 [0.]
 [0.]
 [0.]
 [1.]
 [0.]
 [0.]
 [0.]
 [0.]
 [1.]
 [1.]
 [0.]
 [0.]
 [1.]
 [1.]
 [1.]
 [1.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [1.]
 [0.]
 [0.]
 [1.]
 [0.]
 [0.]
 [1.]
 [0.]
 [1.]
 [1.]
 [1.]
 [1.]
 [0.]
 [0.]
 [0.]
 [1.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.

Se puden ver las salidas, que son parte del comportamiento de las compuertas logicas

## Conclusion

La precision del modelo es de un 79%. Lo que es un puntaje bajo ya que, solo acierta a poco mas de la mitad de los datos con esactitud. Es por esto que este modelo no es recomendado para aplicarlo en la realidad porque entregaria muchos datos errados.Este modelo podria ser utilizado como una aproximacion a la realidad pero teniendo en cuenta que hay un error cercano 34% de errar en el resultado